overlord/hookstate/context.go

```diff
@@ -127,3 +133,57 @@ func (c *Context) Get(key string, value interface{}) error {
 
 	return nil
 }
+
+// State returns the state contained within the context
+func (c *Context) State() *state.State {
+	return c.task.State()
+}
+
+// Cached returns the cached value associated with the provided key.
+// It returns nil if there is no entry for key.
+func (c *Context) Cached(key interface{}) interface{} {
+	cachedCache := c.task.State().Cached(contextCache{})
```

Comment:
Done. I merged all mutex-related operations into context.Lock/Unlock, moving locking responsibility completely out of the class. To make this more palatable I also implemented lock checks.